feat(vhd-cli compare): compare metadata and content of two VHDs

fbeauchamp opened the pull request:
### Check list

> Check if done, if not relevant leave unchecked.

- [ ] PR reference the relevant issue (e.g. `Fixes #007` or `See xoa-support#42`)
- [ ] if UI changes, a screenshot has been added to the PR
- [ ] documentation updated
- `CHANGELOG.unreleased.md`:
  - [ ] enhancement/bug fix entry added
  - [ ] list of packages to release updated (`${name} v${new version}`)
- **I have tested added/updated features** (and impacted code)
  - [ ] unit tests (e.g. [`cron/parse.spec.js`](https://github.com/vatesfr/xen-orchestra/blob/b24400b21de1ebafa1099c56bac1de5c988d9202/%40xen-orchestra/cron/src/parse.spec.js))
  - [ ] if `xo-server` API changes, the corresponding test has been added to/updated on [`xo-server-test`](https://github.com/vatesfr/xen-orchestra/tree/master/packages/xo-server-test)
  - [x] at least manual testing

### Process

1. create a PR as soon as possible
1. mark it as `WiP:` (Work in Progress) if not ready to be merged
1. when you want a review, add a reviewer (and only one)
1. if necessary, update your PR, and re- add a reviewer

From [_the Four Agreements_](https://en.wikipedia.org/wiki/Don_Miguel_Ruiz#The_Four_Agreements):

1. Be impeccable with your word.
1. Don't take anything personally.
1. Don't make assumptions.
1. Always do your best.

fbeauchamp committed:
feat(vhd): implement chunked Vhd

fbeauchamp committed:
feat(vhd-cli): implement copy of a vhd

fbeauchamp committed:
Fix: vhd directory

* Remove commented method
* Rename ChunkVhd, FileVhd and VhdAbstract to VhdDirectory, VhdFile, VhdAbstract
* Refactor ChunkVhd and FileVhd .open to use an object as parameter
* order AbstractVhd properties, fix method and properties visibility
* implement higher level methods to read/write parentLocator
* refactor to compute fullBlockSize, bitmapSize, sectorsOfBitmap and sectorsPerBlocks proeprties on header set
* use getops for help flags in copy
* fix some typos

fbeauchamp committed:
fix: export vhd, assert header, null return

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+      )
+    }
+  }
+}
```
Any reasons for not using `lodash/isEqual`?

julien-f replied:
I guess it makes sense for explicit messages.

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+import { VhdFile, VhdDirectory } from 'vhd-lib'
+
```
Remove this blank line.

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+      } else {
+        throw e
+      }
+    }
```
Move logic in its own function (in this file for now, we'll move it to `Vhd` later).

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+import { resolve } from 'path'
+import Disposable from 'promise-toolbox/Disposable'
+
+const compareObj = (src, dest, path) => {
```
```suggestion
const compareVhd = (src, dest, path) => {
```

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+  for (const key in src) {
+    if (!Object.prototype.hasOwnProperty.call(src, key)) {
+      continue
+    }
```
```suggestion
  for (const key of Object.keys(src)) {
```

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+      same = src[key].equals(dest[key])
+    } else {
+      same = src[key] === dest[key]
+    }
```
```suggestion
    const srcValue = src[key]
    const destValue = dest[key]

    same = srcValue === destValue || (Buffer.isBuffer(srcValue) && Buffer.isBuffer(destValue) && srcValue.equals(destValue))
```

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,111 @@
+
+    if (typeof src[key] === 'object') {
+      compareObj(src[key], dest[key], path + '/' + key)
+      continue
```
IMHO, `continue`, `break` and `returns` complexify the workflow and can lead to maintenance issues, they should be used only when providing clear advantages, otherwise prefer `if`/`else`.

fbeauchamp committed:
fix: use assert.notstrictequals in VhdFile

fbeauchamp committed:
fix fullblock size + code cleanup

fbeauchamp committed:
Fix: ensure parent locator entries have enough space

fbeauchamp committed:
refactor: move header check in the getter

fbeauchamp committed:
fix readParentLocatorData

fbeauchamp committed:
feat: implement vhd-cli

julien-f commented on packages/vhd-lib/src/openVhd.js:
```diff
@@ -0,0 +1,15 @@
+      throw e
+    }
+  }
+  return src
```
```suggestion
  try {
    return await VhdFile.open(handler, path)
  } catch (e) {
    if (e.code !== 'EISDIR') {
      throw e
    }
    return await VhdDirectory.open(handler, path)
  }
```

julien-f commented on packages/vhd-lib/src/openVhd.js:
```diff
@@ -0,0 +1,15 @@
+import { VhdFile, VhdDirectory } from './'
+
+export const openVhd = async (handler, path) => {
```
I see no reasons to use arrow function here.

```suggestion
export async function openVhd (handler, path) {
```

julien-f commented on packages/vhd-lib/src/index.js:
```diff
@@ -10,3 +10,4 @@ export { default as mergeVhd } from './merge'
 export { default as peekFooterFromVhdStream } from './peekFooterFromVhdStream'
 export { VhdDirectory } from './Vhd/VhdDirectory'
 export { VhdFile } from './Vhd/VhdFile'
+export { openVhd } from './openVhd'
```
Keep 'em sorted.

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,82 @@
+      compareVhd(src[key], dest[key], path + '/' + key)
+    } else {
+      const srcValue = src[key]
+      const destValue = dest[key]
```
You can move this at the top of the block.

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,82 @@
+
+const compareVhd = (src, dest, path) => {
+  for (const key of Object.keys(src)) {
+    let same = false
```
I don't think this variable is necessary.

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,82 @@
+      )
+    }
+
+    if (typeof src[key] === 'object') {
```
What of arrays?

fbeauchamp replied:
An array has type object in javascript and Object.keys works with array. Did I miss something?

julien-f replied:
It's not the recommended way, but indeed it should work, fine by me :slightly_smiling_face:

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,82 @@
+import { resolve } from 'path'
+import Disposable from 'promise-toolbox/Disposable'
+
+const compareVhd = (src, dest, path) => {
```
My bad, I previously said to call this function `compareVhd`, but it's not what it does :grimacing:

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,82 @@
+          throw new Error(`Block  ${i} is present in source but not in dest `)
+        }
+      } else {
+        if (dest.containsBlock(i)) {
```
Why not `else if`?

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,82 @@
+      }
+    }
+
+    for (let parentLocatorId = 0; parentLocatorId < 8; parentLocatorId++) {
```
Can we use a constant exported by `vhd-lib` instead of the magic number `8`?

julien-f reviewed (changes requested):
You need to edit `CHANGELOG.unreleased.md` to mark `vhd-lib` and `vhd-cli` as to be released.

fbeauchamp committed:
fix: clean code

fbeauchamp committed:
refactor: parentlocatordata read/write + comment and cleaning

fbeauchamp committed:
Code cleanup

fbeauchamp committed:
feat: export constants from vhd-lib

fbeauchamp committed:
refactor : bat in private

fbeauchamp committed:
refactor : move _getEndOfHeaders and _getEndOfData back to vhdfile

fbeauchamp committed:
feat: add block iterator on vhd

fbeauchamp committed:
rename parent locator id, fix comments

fbeauchamp committed:
feat(vhdfile): reset platformDataOffset if there is no data

fbeauchamp committed:
feat: use the new api for parent locator

fbeauchamp committed:
fix read parent locator and vhddirectory.header read

fbeauchamp committed:
use named constant, make a private getter/setter for blocktable

fbeauchamp committed:
fix: rename chunk file paretnlocatorentry instead of parentlocator, don't expose parentLocator length and space

fbeauchamp committed:
feat: create a directory tree for blocks instead of putting them in the root

fbeauchamp committed:
rework blocks directories

julien-f commented on packages/vhd-lib/src/openVhd.js:
```diff
@@ -0,0 +1,12 @@
+import { VhdFile, VhdDirectory } from './'
+
+export const openVhd = async function (handler, path) {
```
You can use this simpler syntax (also it makes sure that the function is correctly named, regarding of the JS engine):

```suggestion
export async function openVhd(handler, path) {
```

fbeauchamp committed:
Update packages/vhd-lib/src/openVhd.js

Co-authored-by: Julien Fontanet <julien.fontanet@isonoe.net>

julien-f committed:
fix deepCompareObjects

- error message mentionning header
- a buffer is an object, previous code was not handling this correctly
- use type in error message
- don't include buffers in error message

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,80 @@
+    // parent locator entries contains offset that can be different without impacting the vhd
+    // we'll compare them later
+    const { parentLocatorEntry: _, ...srcHeaderWithoutParentLocator } = src.header
+    const { parentLocatorEntry: __, ...destHeaderWithoutParentLocator } = dest.header
```
We should also ignore `tableOffset` and `checksum`.

fbeauchamp committed:
fix: remove more offsets from comparisons

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,80 @@
+    // we'll compare them later
+    const { parentLocatorEntry: srcParentLocatorEntry,tableOffset: srcTableOffset, checksum: srcChecksum,  ...srcHeaderWithoutOffsets } = src.header
+    const { parentLocatorEntry: destParentLocatorEntry,tableOffset: destTableOffset, checksum: destChecksum,  ...destHeaderWithoutOffsets } = dest.header
+    deepCompareObjects(srcHeaderWithoutOffsets, destHeaderWithoutOffsets, 'header')
```
Use `lodash/omit` instead:

```suggestion
    const ignoredEntries= [ 'checksum', 'parentLocatorEntry', 'tableOffset' ]
    deepCompareObjects(
      omit(src.header, ignoredEntries),
      omit(dest.header, ignoredEntries)
    )
```

fbeauchamp committed:
refactor : use omit

julien-f commented on packages/vhd-cli/src/commands/compare.js:
```diff
@@ -0,0 +1,81 @@
+import { openVhd, Constants } from 'vhd-lib'
+import { resolve } from 'path'
+import Disposable from 'promise-toolbox/Disposable'
+import { omit } from 'lodash'
```
You don't need to load the full `lodash`:

```suggestion
import omit from 'lodash/omit'
```

julien-f commented on @xen-orchestra/backups/RemoteAdapter.js:
```diff
@@ -6,7 +6,7 @@ const pDefer = require('promise-toolbox/defer.js')
 const { basename, dirname, join, normalize, resolve } = require('path')
 const { createLogger } = require('@xen-orchestra/log')
-const { createSyntheticStream, mergeVhd, default: Vhd } = require('vhd-lib')
+const { createSyntheticStream, mergeVhd, VhdFile } = require('vhd-lib')
```
Shouldn't it have been changed in the previous PR?

Did it introduce a bug?

fbeauchamp committed:
fix rebase

fbeauchamp committed:
fix : don't import the full lodash
